// periodicjs.ext.user_access_control/controller/userrole.js
'use strict';

var path = require('path'),
    appController = require(path.join(process.cwd(),'app/controller/application')),
    applicationController,
    appSettings,
    mongoose,
    Userrole,
    Userprivilege,
    // Usergroup,
    logger;

var index = function(req, res){
    applicationController.getPluginViewDefaultTemplate(
        {
            viewname:'p-admin/userrole/index',
            themefileext:appSettings.templatefileextension,
            extname: 'periodicjs.ext.admin'
        },
        function(err,templatepath){
            applicationController.handleDocumentQueryRender({
                res:res,
                req:req,
                renderView:templatepath,
                responseData:{
                    pagedata:{
                        title:'Manage Users Roles',
                        extensions:applicationController.getAdminMenu()
                    },
                    userroles: req.controllerData.userroles,
                    user:req.user
                }
            });
        }
    );
};

var create = function (req, res) {
    var newuserrole = applicationController.removeEmptyObjectValues(req.body);
    newuserrole.name = applicationController.makeNiceName(newuserrole.title);
    newuserrole.author = [req.user._id];

    // console.log(newitem);
    applicationController.createModel({
        model: Userrole,
        newdoc: newuserrole,
        res: res,
        req: req,
        successredirect: '/p-admin/userrole/edit/',
        appendid: true
    });
};

var userrole_new = function(req, res){
    applicationController.getPluginViewDefaultTemplate(
        {
            viewname:'p-admin/userrole/new',
            themefileext:appSettings.templatefileextension,
            extname: 'periodicjs.ext.admin'
        },
        function(err,templatepath){
            applicationController.handleDocumentQueryRender({
                res:res,
                req:req,
                renderView:templatepath,
                responseData:{
                    pagedata:{
                        title:'New User Role',
                        headerjs: ['/extensions/periodicjs.ext.admin/js/useraccess.min.js'],
                        extensions:applicationController.getAdminMenu()
                    },
                    userrole:null, //req.controllerData.userrole
                    user:req.user
                }
            });
        }
    );
};

var controller = function(resources){
    logger = resources.logger;
    mongoose = resources.mongoose;
    appSettings = resources.settings;
    applicationController = new appController(resources);
    Userrole = mongoose.model('Userrole');
    Userprivilege = mongoose.model('Userprivilege');
    // Usergroup = mongoose.model('Usergroup');

    return{
        index:index,
        create:create,
        userrole_new:userrole_new
    };
};

module.exports = controller;
