// periodicjs.ext.user_access_control/controller/userrole.js
'use strict';

var path = require('path'),
    appController = require(path.join(process.cwd(),'app/controller/application')),
    applicationController,
    appSettings,
    mongoose,
    Userrole,
    Userprivilege,
    logger;

var index = function(req, res){
    applicationController.getPluginViewDefaultTemplate(
        {
            viewname:'p-admin/userroles/index',
            themefileext:appSettings.templatefileextension,
            extname: 'periodicjs.ext.admin'
        },
        function(err,templatepath){
            applicationController.handleDocumentQueryRender({
                res:res,
                req:req,
                renderView:templatepath,
                responseData:{
                    pagedata:{
                        title:'Manage Users Roles',
                        extensions:applicationController.getAdminMenu()
                    },
                    userroles: req.controllerData.userroles,
                    user:req.user
                }
            });
        }
    );
};

var create = function (req, res) {
    var newuserrole = applicationController.removeEmptyObjectValues(req.body);
    newuserrole.name = applicationController.makeNiceName(newuserrole.title);
    newuserrole.author = [req.user._id];

    // console.log(newitem);
    applicationController.createModel({
        model: Userrole,
        newdoc: newuserrole,
        res: res,
        req: req,
        successredirect: '/p-admin/userrole/edit/',
        appendid: true
    });
};

var userrole_new = function(req, res){
    applicationController.getPluginViewDefaultTemplate(
        {
            viewname:'p-admin/userroles/new',
            themefileext:appSettings.templatefileextension,
            extname: 'periodicjs.ext.admin'
        },
        function(err,templatepath){
            applicationController.handleDocumentQueryRender({
                res:res,
                req:req,
                renderView:templatepath,
                responseData:{
                    pagedata:{
                        title:'New User Role',
                        headerjs: ['/extensions/periodicjs.ext.admin/js/useraccess.min.js'],
                        extensions:applicationController.getAdminMenu()
                    },
                    userrole:null, //req.controllerData.userrole
                    user:req.user
                }
            });
        }
    );
};

var show = function (req, res) {
    applicationController.getPluginViewDefaultTemplate({
            viewname:'p-admin/userroles/edit',
            themefileext:appSettings.templatefileextension,
            extname: 'periodicjs.ext.admin'
        },
        function (err, templatepath) {
            applicationController.handleDocumentQueryRender({
                res: res,
                req: req,
                renderView: templatepath,
                responseData: {
                    pagedata: {
                        title: req.controllerData.userrole.title,
                        headerjs: ['/extensions/periodicjs.ext.admin/js/useraccess.min.js'],
                        extensions:applicationController.getAdminMenu()
                    },
                    userrole:req.controllerData.userrole,
                    user: req.user
                }
            });
        }
    );
};

var update = function (req, res) {
    var updateuserrole = applicationController.removeEmptyObjectValues(req.body);

    updateuserrole.name = applicationController.makeNiceName(updateuserrole.title);

    applicationController.updateModel({
        model: Userrole,
        id: updateuserrole.docid,
        updatedoc: updateuserrole,
        saverevision: true,
        population: 'privileges',
        res: res,
        req: req,
        successredirect: '/p-admin/userrole/edit/',
        appendid: true
    });
};

var remove = function (req, res) {
    var userrole = req.controllerData.userrole;
    applicationController.deleteModel({
        model: Userrole,
        deleteid: userrole._id,
        req: req,
        res: res,
        callback: function(err){
             if (err) {
                applicationController.handleDocumentQueryErrorResponse({
                    err: err,
                    res: res,
                    req: req
                });
            }
            else {
                applicationController.handleDocumentQueryRender({
                    req: req,
                    res: res,
                    redirecturl: '/p-admin/userroles',
                    responseData: {
                        result: "success",
                        data: "deleted"
                    }
                });
            }
        }
    });
};

var controller = function(resources){
    logger = resources.logger;
    mongoose = resources.mongoose;
    appSettings = resources.settings;
    applicationController = new appController(resources);
    Userrole = mongoose.model('Userrole');
    Userprivilege = mongoose.model('Userprivilege');
    // Usergroup = mongoose.model('Usergroup');

    return{
        index:index,
        create:create,
        userrole_new:userrole_new,
        show:show,
        update:update,
        remove:remove
    };
};

module.exports = controller;
